// @prisma/engine-core/dist/undici.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Undici = void 0;
const get_stream_1 = __importDefault(require("get-stream"));
const undici_1 = require("undici");
class Undici {
    constructor(url, moreArgs) {
        this.closed = false;
        this.pool = new undici_1.Pool(url, {
            connections: 100,
            pipelining: 10,
            requestTimeout: 0,
            socketTimeout: 0,
            ...moreArgs,
        });
    }
    request(body, customHeaders) {
        return new Promise((resolve, reject) => {
            this.pool.request({
                path: '/',
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    ...customHeaders,
                },
                body,
            }, async (err, result) => {
                if (err) {
                    reject(err);
                }
                else {
                    const { statusCode, headers, body } = result;
                    const data = JSON.parse(await get_stream_1.default(body));
                    resolve({ statusCode, headers, data });
                }
            });
        });
    }
    status() {
        return new Promise((resolve, reject) => {
            this.pool.request({
                path: '/',
                method: 'GET',
            }, async (err, result) => {
                if (err) {
                    reject(err);
                }
                else {
                    const { statusCode, headers, body } = result;
                    const data = JSON.parse(await get_stream_1.default(body));
                    resolve({ statusCode, headers, data });
                }
            });
        });
    }
    close() {
        if (!this.closed) {
            this.pool.close(() => {
                // ignore close error
            });
        }
        this.closed = true;
    }
}
exports.Undici = Undici;
